core/modules/wdb/ChunkResource.cc

```diff
@@ -235,8 +279,112 @@ class Backend {
         }
     }
 
+    /// Run the 'query'. If it fails, terminate the program.
+    void _execLockSql(std::string const& query) {
+        LOGF_DEBUG("execLockSql %1%" % query);
+        sql::SqlErrorObject err;
+        if(!_sqlConn.runQuery(query, err)) {
+            LOGF_ERROR("Lock failed, exiting. query=%1%  err=%2%" % query % err.printErrMsg());
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    /// Return the status of the lock on the in memory tables.
+    LockStatus _memLockStatus() {
+        std::string sql = "SELECT uid FROM " + _lockDbTbl + " WHERE keyId = 1;";
+        sql::SqlResults results;
+        sql::SqlErrorObject err;
+        if (!_sqlConn.runQuery(sql, results, err)) {
+            // Assuming UNLOCKED should be safe as either it must be LOCKED_OURS to continue
+            // or we are about to try to lock. Failure to lock will cause the program to exit.
+            std::ostringstream emsg;
+            emsg << "memLockStatus query failed, assuming UNLOCKED. " << sql << " err=" << err.printErrMsg();
+            LOGF_WARN("%1%" % emsg.str());
+            return UNLOCKED;
+        }
+        std::string uidStr;
+        if (!results.extractFirstValue(uidStr, err)) {
+            std::ostringstream emsg;
+            emsg << "memLockStatus unexpected results, assuming LOCKED_OTHER. err=" << err.printErrMsg();
+            LOGF_WARN("%1%" % emsg.str());
+            return LOCKED_OTHER;
+        }
+        int uid = atoi(uidStr.c_str());
+        if (uid != _uid) {
+            std::ostringstream emsg;
+            emsg << "memLockStatus LOCKED_OTHER wrong uid. Expected " << _uid << " got " << uid
+                << " err=" << err.printErrMsg();
+            LOGF_WARN("%1%" % emsg.str());
+            return LOCKED_OTHER;
+        }
+        return LOCKED_OURS;
+    }
+
+    /// Take the lock on the in memory tables, even if held by another.
+    // Terminate the program if we can't take the lock.
+    // This must be run before any other operations on in memory tables.
+    void _memLockLock() {
```

Comment:
Actually, maybe it would be better to just get rid of the explicit release code entirely (sidestepping Andy's concern about the Backend destructor) and note that if the worker exits, you must also restart mysqld.